chain/beacon/chain.go

```diff
@@ -217,15 +218,72 @@ func (c *chainStore) shouldSync(last *chain.Beacon, newB likeBeacon) bool {
 	return newB.GetRound() > last.GetRound()+1
 }
 
-// RunSync will sync up with other nodes and fill the store. If upTo is equal to
-// 0, then it will follow the chain indefinitely. If peers is nil, it uses the
-// peers of the current group.
-func (c *chainStore) RunSync(upTo uint64, peers []net.Peer) {
-	if peers == nil {
+// RunSync will sync up with other nodes and fill the store. If from is negative, it
+// will start from the latest stored beacon. If upTo is equal to 0, then it
+// will follow the chain indefinitely. If peers is nil, it uses the peers of
+// the current group.
+func (c *chainStore) RunSync(from, upTo uint64, peers []net.Peer) {
+	if len(peers) == 0 {
 		peers = toPeers(c.crypto.GetGroup().Nodes)
 	}
 
-	c.syncm.RequestSync(peers, upTo)
+	c.syncm.RequestSync(from, upTo, peers)
+}
+
+// ValidateChain will validate the chain in the chain store up to the given beacon.
+func (c *chainStore) ValidateChain(ctx context.Context, upTo uint64, cb func(r, u uint64)) ([]uint64, error) {
+	logger := c.l.Named("pastBeaconChecker")
+	logger.Debugw("Starting to check past beacons", "upTo", upTo)
+
+	last, err := c.Last()
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch and check last beacon in store: %w", err)
+	}
+
+	if last.Round < upTo {
+		logger.Errorw("No beacon stored above", "last round", last.Round, "requested round", upTo)
+		logger.Infow("Checking beacons only up to the last stored", "round", last.Round)
+		upTo = last.Round
+	}
+
+	var faultyBeacons []uint64
+	// notice that we do not validate the genesis round 0
+	for i := 1; i < c.Len(); i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+		// we call our callback with the round to send the progress, N.B. we need to do it before returning
+		cb(uint64(i), upTo)
```

Comment:
sending progress a million times may end up being more expensive than anything else. I wonder if we want to callback every 1000 or something like that?